### PR TITLE
[WIP][System.Reflection] CoreFX import for FieldInfo type.

### PR DESCRIPTION
**Pull request by MaximLipnin**
Part of #9660.

The changes:
- included FieldInfo type from CoreFX + related xunit tests;
- made Mono-implemented class partial; removed duplicated code;
- moved GetOptionalCustomModifiers and GetRequiredCustomModifiers methods to MonoField type;
- moved GetTypeModifiers extern method and related icall to MonoField type;
- updated ToString method in CustomAttributeTypedArgument type with CoreFX implementation;

### CODE REVIEW

**Comment by MaximLipnin** [comment]
@monojenkins build deb with monolite

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs [hunk 1]
```diff
@@ -47,7 +49,7 @@ public struct CustomAttributeTypedArgument {
 				throw new ArgumentNullException ("argumentType");
 
 			this.argumentType = argumentType;
-			this.value = value;
+			this.value = (value == null) ? null : CanonicalizeValue (value);
 
 			// MS seems to convert arrays into a ReadOnlyCollection
 			if (value is Array) {
```
[the remark this answers is not in the document]
Comment (MaximLipnin) [reply mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: Not sure, I've updated only some problem part here. Anyway CustomAttributeTypedArgument type will be replaced with corert stuff on some of next steps.

File: mcs/class/corlib/System.Reflection/MonoField.cs [hunk 2]
```diff
@@ -329,5 +329,22 @@ static int get_core_clr_security_level ()
 		
 		[MethodImplAttribute (MethodImplOptions.InternalCall)]
 		internal static extern int get_metadata_token (MonoField monoField);
+
+		[MethodImplAttribute (MethodImplOptions.InternalCall)]
+		extern Type[] GetTypeModifiers (bool optional);
+
+		public override Type[] GetOptionalCustomModifiers () {
+			Type[] types = GetTypeModifiers (true);
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/MonoField.cs]: Nicer syntax would be `=> GetTypeModifiers (true) ?? Type.EmptyTypes`

File: mcs/class/corlib/System.Reflection/FieldInfo.cs [hunk 3]
```diff
@@ -164,33 +59,11 @@ public static FieldInfo GetFieldFromHandle (RuntimeFieldHandle handle, RuntimeTy
...
-		// Note: making this abstract imposes an implementation requirement
-		//       on any class that derives from it.  However, since it's also
-		//       internal, that means only classes inside corlib can derive
-		//       from FieldInfo.  See
-		//
-		//          errors/cs0534-4.cs errors/CS0534-4-lib.cs
-		//
-		//          class/Microsoft.JScript/Microsoft.JScript/JSFieldInfo.cs
-		//
 		internal virtual int GetFieldOffset ()
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/FieldInfo.cs]: Could we make it abstract ?

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs [hunk 4]
```diff
@@ -47,7 +49,7 @@ public struct CustomAttributeTypedArgument {
 				throw new ArgumentNullException ("argumentType");
 
 			this.argumentType = argumentType;
-			this.value = value;
+			this.value = (value == null) ? null : CanonicalizeValue (value);
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: What about setting the value only when it's not null?

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs [hunk 5]
```diff
@@ -82,17 +84,46 @@ public CustomAttributeTypedArgument (object value)
...
-
-			return val;
+			if (ArgumentType == null)
+				return base.ToString (); // Someone called ToString() on "default(CustomAttributeTypedArgument)"
+
+			try	{
+				if (ArgumentType.IsEnum)
+					return string.Format (CultureInfo.CurrentCulture, typed ? "{0}" : "({1}){0}", Value, ArgumentType.FullName);
+
+				else if (Value == null)
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: else(s) are redundant

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs [hunk 6]
```diff
@@ -82,17 +84,46 @@ public CustomAttributeTypedArgument (object value)
...
+					return string.Format (CultureInfo.CurrentCulture, "\"{0}\"", Value);
+
+				else if (ArgumentType == typeof (char))
+					return string.Format (CultureInfo.CurrentCulture, "'{0}'", Value);
+
+				else if (ArgumentType == typeof (Type))
+					return string.Format (CultureInfo.CurrentCulture, "typeof({0})", ((Type)Value).FullName);
+
+				else if (ArgumentType.IsArray) {
+					string result = null;
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: This should use StringBuilder

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs [hunk 7]
```diff
@@ -118,8 +149,12 @@ public override int GetHashCode ()
 		{
 			return !left.Equals (right);
 		}
-	}
 
+		private static object CanonicalizeValue (object value)
+		{
+			if (value.GetType ().IsEnum)
```
Comment (marek-safar) [inline mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: Perhaps `if (value is Enum e) return e.GetValue ();` would be better

**Comment by MaximLipnin** [comment]
@marek-safar Should the comments regarding changes in `CustomAttributeTypedArgument.cs` be addressed here? It'll come back again with corert version.

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs (hunk 4 above)
Comment (MaximLipnin) [reply mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: In this case I get `error CS0171: Field 'CustomAttributeTypedArgument.value' must be fully assigned before control is returned to the caller`.

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs (hunk 5 above)
Comment (MaximLipnin) [reply mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: Not sure why

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs (hunk 4 above)
Comment (marek-safar) [reply mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: oh, this is a struct (nevermind then)

File: mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs (hunk 5 above)
Comment (marek-safar) [reply mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: You have returns before

Comment (MaximLipnin) [reply mcs/class/corlib/System.Reflection/CustomAttributeTypedArgument.cs]: ok, I thought you meant the all `else` branches

**Comment by MaximLipnin** [comment]
@monojenkins build failed

**Comment by MaximLipnin** [comment]
There are no build issues at the moment; related pull request to corefx fork is not critical for merging and can be done later.

**Comment by akoeplinger** [comment]
@monojenkins commit apidiff and csproj

**Comment by MaximLipnin** [comment]
Thanks for review!

**Pull request merged**